src/longrange_force_types/force_ewald_cabana_neigh_impl.h

```diff
@@ -0,0 +1,597 @@
+
+  //This seems awkward. Could remove?
+  Kokkos::View<double *, MemorySpace> domain_length( "domain length", 6 );
+  domain_length( 0 ) = system->sub_domain_lo_x;
+  domain_length( 1 ) = system->sub_domain_hi_x;
+  domain_length( 2 ) = system->sub_domain_lo_y;
+  domain_length( 3 ) = system->sub_domain_hi_y;
+  domain_length( 4 ) = system->sub_domain_lo_z;
+  domain_length( 5 ) = system->sub_domain_hi_z;
+
+  // compute subdomain size and make it available in the kernels
+  Kokkos::View<double *, MemorySpace> sub_domain_size( "sub_domain size", 3);
+  sub_domain_size( 0 ) = system->sub_domain_x;
+  sub_domain_size( 1 ) = system->sub_domain_y;
+  sub_domain_size( 2 ) = system->sub_domain_z;
+  //sub_domain_size( 0 ) = (system->sub_domain_hi_x + system->sub_domain_x) - (system_sub_domain_lo_x - system_sub_domain_x);
+  //sub_domain_size( 1 ) = (system->sub_domain_hi_y + system->sub_domain_y) - (system_sub_domain_lo_y - system_sub_domain_y);
+  //sub_domain_size( 2 ) = (system->sub_domain_hi_z + system->sub_domain_z) - (system_sub_domain_lo_z - system_sub_domain_z);
+  //sub_domain_size( 1 ) = system->subdomain_hi_y - system_subdomain_y;
+  //sub_domain_size( 2 ) = system->subdomain_hi_z - system_subdomain_z;
+ 
+  // compute domain size and make it available in the kernels
+  Kokkos::View<double *, MemorySpace> domain_size( "domain size", 3 );
+  domain_size( 0 ) = system->domain_x;//TODO: Seems very nnecessary. Remove
+  domain_size( 1 ) = system->domain_y;
+  domain_size( 2 ) = system->domain_z;
+  //domain_size( 0 ) = (system->domain_hi_x + system->domain_x) - (system->domain_lo_x - system->domain_x);
+  //domain_size( 1 ) = (system->domain_hi_y + system->domain_y) - (system->domain_lo_y - system->domain_y);
+  //domain_size( 2 ) = (system->domain_hi_z + system->domain_z) - (system->domain_lo_z - system->domain_z);
+ 
+  //get the solver parameters
+  double alpha = _alpha;
+  double r_max = _r_max;
+  double eps_r = _eps_r;
+  double k_max = _k_max;
+
+  // store MPI information
+  int rank, n_ranks;
+  std::vector<int> loc_dims( 3 );
+  std::vector<int> cart_dims( 3 );
+  std::vector<int> cart_periods( 3 );
+  MPI_Comm_rank( comm, &rank );
+  MPI_Comm_size( comm, &n_ranks );
+  MPI_Cart_get( comm, 3, cart_dims.data(), cart_periods.data(),
+                loc_dims.data() );
+
+  // neighbor information
+  std::vector<int> neighbor_low( 3 );
+  std::vector<int> neighbor_up( 3 );
+
+  // get neighbors in parallel decomposition
+  for ( int dim = 0; dim < 3; ++dim )
+  {
+      MPI_Cart_shift( comm, dim, 1, &neighbor_low.at( dim ),
+                      &neighbor_up.at( dim ) );
+  }
+
+  // initialize potential and force to zero
+  auto init_parameters = KOKKOS_LAMBDA( const int idx )
+    {
+        p( idx ) = 0.0;
+        f( idx, 0 ) = 0.0;
+        f( idx, 1 ) = 0.0;
+        f( idx, 2 ) = 0.0;
+    }; 
+  Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>( 0, N_max ), init_parameters );
```

Comment:
I think you want to be using `N_local` (or potentially `N_local+N_ghost`), `N_max` is the max allocated size of the AoSoA

Reply:
Yes, I think so? @rhalver I think I am mixing up your use of `n_max` and the `System->N_Max` used here. Could you comment on this?